src/EventListener/VisitorTrackingSubscriber.php

```diff
@@ -0,0 +1,226 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Alpha\VisitorTrackingBundle\EventListener;
+
+use Alpha\VisitorTrackingBundle\Entity\Lifetime;
+use Alpha\VisitorTrackingBundle\Entity\PageView;
+use Alpha\VisitorTrackingBundle\Entity\Session;
+use Alpha\VisitorTrackingBundle\Storage\SessionStore;
+use Doctrine\Common\Inflector\Inflector;
+use Doctrine\ORM\EntityManager;
+use Symfony\Bundle\SecurityBundle\Security\FirewallMap;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpFoundation\Cookie;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\HttpKernel\Event\FilterResponseEvent;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+
+/**
+ * Tracks the source of a session and each page view in that session.
+ */
+class VisitorTrackingSubscriber implements EventSubscriberInterface
+{
+    public const COOKIE_LIFETIME = 'lifetime';
+
+    public const COOKIE_SESSION = 'session';
+
+    protected const UTM_CODES = [
+        'utm_source' => 'utm_source',
+        'utm_medium' => 'utm_medium',
+        'utm_campaign' => 'utm_campaign',
+        'utm_term' => 'utm_term',
+        'utm_content' => 'utm_content',
+    ];
+
+    protected const COOKIE_SESSION_TTL = '+2 years';
+
+    /**
+     * @var EntityManager
+     */
+    protected $em;
+
+    /**
+     * @var SessionStore
+     */
+    private $sessionStore;
+
+    /**
+     * @var array
+     */
+    private $firewallBlacklist;
+
+    /**
+     * @var FirewallMap
+     */
+    private $firewallMap;
+
+    public function __construct(EntityManager $em, SessionStore $sessionStore, array $firewallBlacklist, FirewallMap $firewallMap)
```

Comment:
We can improve the first argument's injection to be for `EntityManagerInterface`.

Reply:
We can't.

It's reliant on the `EntityManager`'s signature allowing to specify a single entity to flush. 

Dropping this behaviour is possible, uncertain if it's a good idea at present. Doctrine wants to the drop this functionality entirely however, so it's certainly something worth looking at, as we'll have to change this at some point in the future.

Reply:
Oh, I've missed that point. Good call!